doc/configuration.rst
Configuration
=============

In this section you will learn how to configure your PEER instance. As any
Django project all configuration options are stored in a ``settings.py`` file,
which is located inside the main PEER directory.

However it is not recommended to change the ``settings.py`` file directly since
that would make upgrades more difficult for you. You should see this file
as the file with the **default** configuration settings. You should put
your own settings in a file called ``local_settings.py``. The options you
put there will override the same options located in the ``settings.py`` file.

First we will cover the options that you **must** change in order to run
the PEER instance correctly. Then we will see other optional settings that
you may want to change but that are ok with their default values. At the
end of this section you will see ``local_settings.py`` example, also you can
find at the root of the project a ``local_settings.example`` file.

If you change any setting you must restart your web server in order to
see your new settings applied.

Settings you should change
--------------------------

Database settings
~~~~~~~~~~~~~~~~~

Here you tell PEER the necessary information to connect to your database. The
default values will not probably fit your case:

.. code-block:: python

 DATABASES = {
     'default': {
         'ENGINE': 'django.db.backends.postgresql_psycopg2',
         'NAME': 'peer',
         'USER': 'peer',
         'PASSWORD': 'peer',
         'HOST': '',
         'PORT': '',
     }
 }

This option is pretty much self explanatory. Check the Django documentation
to learn all the details about this setting.

The Site object
~~~~~~~~~~~~~~~

Django uses a Site model to store information about the domain your project
is running at. This is used when composing links that are not shown in the
web application itself, such as links in the emails that are sent.

So you need to log in as the administrator user and go to the `Django admin
interface`_ to edit the only Site object you should have:

.. _`Django admin interface`: http://127.0.0.1:8000/admin/

.. figure:: _static/change-site-object.png

It is very important to enter the correct Domain name so users will be able
to complete the registration process by clicking in correct links in their
inboxes.

.. note::
  The id of your Site object should match the value of the SITE_ID setting.
  You shouldn't need to do anything special here since this is the default
  behaviour.

Mail settings
~~~~~~~~~~~~~

The PEER instance needs to send email in order to correctly perform several
actions such as sending registration activation links or warning users when
their metadata is about to expire.

The *DEFAULT_FROM_EMAIL* settings is used as the *From* header in all these
emails:

.. code-block:: python

 DEFAULT_FROM_EMAIL = 'no-reply@example.com'

You will probably want to change this email address to fit the subdomain
where you installed the PEER instance.

The next settings define where is your SMTP server and how to connect to it:

.. code-block:: python

 EMAIL_HOST = 'smtp.example.com'
 EMAIL_PORT = 25

.. note::
  You can check Django settings reference if your SMTP server require
  authentication.

Recaptcha keys and the Django SECRET_KEY
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

PEER uses the Recaptcha service to prevent spammers to register into your
PEER instance. In order to use such service you should register your PEER
site into Recaptcha and they will give you a public and a private key. You
have to write these keys in your ``local_settings.py`` file:

.. code-block:: python

 RECAPTCHA_PUBLIC_KEY = ''
 RECAPTCHA_PRIVATE_KEY = ''

As you can see the default values for these options are empty so you must
not skip this step if you want the user registration to operate correctly.

Django uses another key for some features. This key is called the
*SECRET_KEY* and should be unique. When a new Django project is created the
django-admin.py program will create such a key for you but the PEER project
is already cloned so you will need to create it yourself. One way to create
such a key is executing the following command:

.. code-block:: bash

 python -c "from random import choice; print ''.join([choice('abcdefghijklmnopqrstuvwxyz0123456789\!@#$%^&*(-_=+)') for i in range(50)])"

And now paste it into the SECRET_KEY option:

.. code-block:: python

 SECRET_KEY = ''

As before, the default value of the SECRET_KEY is empty for security reasons.

MEDIA_ROOT setting
~~~~~~~~~~~~~~~~~~

This settings specifies a directory where user files will be stored. In PEER,
this directory contains the main metadata repository, usually stored as a
GIT repository.

.. code-block:: python

 MEDIA_ROOT = os.path.join(BASEDIR, 'media')

By default the *MEDIA_ROOT* root lives inside the PEER main directory which
is **not** what you usually want. You really want to put this directory
outside your PEER main directory in order to prevent data lost when
upgrading the PEER software. Actually the upgrade process does not lost any
data but as the new version is installed into another directory you will
need to move the media root or change the MEDIA_ROOT setting anytime you
upgrade PEER.

Also, remember that the user that your web server is run as needs to have
write access to the *MEDIA_ROOT* directory.

Version control for SAML metadata documents
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In this system, endpoints are represented by SAML documents,
that are kept under version control by `GIT`_. Some other VCS might be
used; this is controlled by the setting:

.. _`GIT`: http://git-scm.com

.. code-block:: python

 VFF_BACKEND = 'vff.git_backend.GitBackend'

This setting is required, and, at the moment, there is only one VCS backend,
the GIT backend, so this setting must be set exactly as shown.

The git backend can be configured by 2 additional settings, both of which are
optional. With *VFF_REPO_ROOT* you point to the git repo that will keep the
documents. This repo can exist previously, otherwise it will be initialized
on first access. If this setting is not provided, the system will default to
``os.path.join(settings.MEDIA_ROOT, 'vf_repo``:

.. code-block:: python

 VFF_REPO_ROOT = '/path/to/my/repo'

With *VFF_REPO_PATH*, you can specify a relative path within the repository.
This may be useful if you use a repository that was in use before for other
purposes, and want to keep metadata files separated from other content. If
unset, it will default to no subpath:

.. code-block:: python

 VFF_REPO_PATH = 'my/subpath'

Terms of use texts
~~~~~~~~~~~~~~~~~~

PEER asks the user to accept some terms of use in two situations:

- When the user is registered in the site.
- When the user import metadata from his computer or a remote location.

In both cases a legal text should be presented to the user in order to
prevent future complications. These text must be stored in UTF-8 encoded
text files and you will configure their locations with a couple of settings:

.. code-block:: python

 USER_REGISTER_TERMS_OF_USE = os.path.join(BASEDIR, 'user_register_terms_of_use.txt')
 METADATA_IMPORT_TERMS_OF_USE = os.path.join(BASEDIR, 'metadata_import_terms_of_use.txt')

The default values reference a couple of files located inside the PEER main
directory. The content of these files is not very useful skeaking about legal
terms, so you should contact your lawyers and create your own files. As with
the `MEDIA_ROOT setting`_, it is recommended to place these files outside
the PEER main directory to make upgrades easier.

Settings you may change
-----------------------

Theme
~~~~~

PEER look can be customizable by changing the PEER_THEME setting:

.. code-block:: python

 PEER_THEME = {
     'LINK_COLOR': '#5669CE',
     'LINK_HOVER': '#1631BC',
     'HEADING_COLOR': '#1631BC',
     'INDEX_HEADING_COLOR': '#ff7b33',
     'HEADER_BACKGROUND': '',
     'CONTENT_BACKGROUND': '',
     'FOOTER_BACKGROUND': '',
     'HOME_TITLE': 'Nice to meet you!!',
     'HOME_SUBTITLE': 'Say hello to federated worldwide services',
     'JQUERY_UI_THEME': 'default-theme',
 }

Each element of this dictionary has its own purpose:

LINK_COLOR
 Foreground color for the links.

LINK_HOVER
 Foreground color for he links when the mouse cursor is over hem.

HEADING_COLOR:
 Foreground color for the heading section.

INDEX_HEADING_COLOR:
 Foreground color for the heading section of the index page (homepage).

HEADER_BACKGROUND:
 Background color for the header section.

CONTENT_BACKGROUND:
 Background color for the content section.

HOME_TITLE:
 Text that will be shown in the main banner of the homepage.

HOME_SUBTITLE:
 Secondary text that will be shown in the main banner of the homepage.

JQUERY_UI_THEME:
 jQuery UI theme to use. You can generate these themes using the `Theme
 Roller application`_. Then you should put this theme inside the css
 directory. This theme should be for jQuery UI 1.8.14 version. This theme
 will affect the look of every button, icon, user messages and other widgets
 accross the whole site.

.. _`Theme Roller application`: http://jqueryui.com/themeroller/

Check the :doc:`branding` section for more information about how to
create a coherent theme.


Registration settings
~~~~~~~~~~~~~~~~~~~~~

These settings affect the user registration process. Right one only one
setting exists:

.. code-block:: python

 ACCOUNT_ACTIVATION_DAYS = 2

This is the number of days the activation key can be used. After this period,
the user will need to register again.

Metadata Validation
~~~~~~~~~~~~~~~~~~~

The METADATA_VALIDATORS settings specifies the validators that will be used
in the validation process that happens every time an entity's metadata is
changed. It is a list of strings, each string representing the full path
of a python function, that is the validator:

.. code-block:: python

 METADATA_VALIDATORS = (
     'peer.entity.validation.validate_xml_syntax',
     'peer.entity.validation.validate_domain_in_endpoints',
     'peer.entity.validation.validate_metadata_permissions',
 )

In order to save the changes of an entity's metadata all the validators must
succeeded.

A validator is just a single python function with the following interface:

* It receives two arguments: the entity object and an XML string representing
  the metadata.
* It returns a list of error messages or an empty list if the XML string is
  valid.

Check the provided validators for examples about how to write your own
validators.

Metadata Permissions
~~~~~~~~~~~~~~~~~~~~

The METADATA_PERMISSIONS settings specified the SAML metadata elements whose
permissions can be managed. The permissions for each element are for adding,
deleting and modifying the element. Any XPATH not specified in the settings file
will by default will give permission any user to do anything. The format of each
element of the settings is:

.. code-block:: python

 MEATADATA_PERMISSIONS = (
      ('XPATH', 'permission_name', 'Permission Description')
 )

In a concrete example:

.. code-block:: python

 METADATA_PERMISSIONS = (
     ('.//md:ServiceDescription', 'service_descriptor', 'Service Description'),
     ('.//mdui:Description', 'description', 'Description'),
     ('.//md:OrganizationName', 'organization_name', 'Organization Name'),
 )

Once the pemissions are specified a special migration is needed:

.. code-block:: bash

 $ django-admin.py migrate --all

In order to manage permissions for a given user, you need to login as superuser
in the `Django admin interface`_, browse to *Users* and pick the user whose
permissions you want to change (by default, once a permission is added to the
settings, a normal user won't be able to do anything in the attribute specified
in the settings). In *User permissions* there is a panel with the permissions that are
available. For each element specified in the settings the permissions for
adding, deleting and modifying should be present in the panel*; i.e.:
permissions *Can add <Permission Description>*, *Can edit <Permission
Description>* and *Can delete <Permission Description>*. To give the user a
permission, pick the permission and move it to *Chosen user permissions*.

.. figure:: _static/grant_permissions.png

Any SAML metadata element not present in the settings has its permissions
enabled by default. Once it's added a normal user won't be able to add, modify
or delete any SAML medata element unless is not present in the settings and a
superuser has granted the user with the permissions.

.. note::

   As of version 0.7.0 of PEER, only the XPATH subset supported by the `XPATH
   ElementTree API`_ can be used. This means that selectors by element value,
   for example, are not supported.

.. _`XPATH ElementTree API`: http://effbot.org/zone/element-xpath.htm


SAMLmetaJS plugins
~~~~~~~~~~~~~~~~~~

SAMLmetaJS is a jQuery plugin that turns a simple HTML textarea element into
a full blown SAML metadata editor. It has a small core and several plugins
for editing specific parts of the metadata XML.

With this setting you can set which plugins are going to be active and in
which order. This will affect the tabs that appear in the metadata edition
view.

.. code-block:: python

 SAML_META_JS_PLUGINS = ('info', 'org', 'contact', 'saml2sp', 'certs',
                         'attributes')

Check the `SAMLmetaJS website`_ for a complete list of all available plugins.

.. _`SAMLmetaJS website`: http://samlmetajs.simplesamlphp.org/

Pagination and feeds settings
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

With these settings you can control the number of entities that are
shown in certain circunstances.

The MAX_FEED_ENTRIES setting controls the number of entities that are
returned in the global rss feed.

.. code-block:: python

 MAX_FEED_ENTRIES = 10

If you do not define this setting the global rss feed will return the full
set of entities. Be careful if you have a lot of entities since this can
be degradate performance.

The ENTITIES_PER_PAGE setting controls the number of entities that
are displayed in each page of the full list view and the search results view.

.. code-block:: python

 ENTITIES_PER_PAGE = 10

Expiration warning
~~~~~~~~~~~~~~~~~~

The EXPIRATION_WARNING_TIMEDELTA setting specifies the time threshold that
should be used to determine if a warning email should be sent when the
metadata of an entity is about to expire. If the time when the metadata
is expired minus the EXPIRATION_WARNING_TIMEDELTA is greater than the current
time, a warning email is sent to the entity's team. For example, if the
metadata expires the 17th of September of 2011 at 16:00 and the
EXPIRATION_WARNING_TIMEDELTA is set to 5 hours, that day at 11:00  a warning
email will be sent.

The value of this settings should be a datetime.timedelta object. Check the
`Python documentation`_ for valid units for this object.

.. _`Python documentation`: http://docs.python.org/library/datetime.html#timedelta-objects

.. code-block:: python

 EXPIRATION_WARNING_TIMEDELTA = datetime.timedelta(days=1)

This feature requires that you setup a cron job that calls the
*expirationwarnings* PEER command. Something like this should work:

.. code-block:: bash

 0 * * * * /var/www/peer/bin/django-admin.py expirationwarnings --settings=peer.settings

Metadata refresh
~~~~~~~~~~~~~~~~

In order for the users to use the metarefresh option in :ref:`entity-management`,
this cron job needs to be setup:

.. code-block:: bash

 * * * * * /var/www/peer/bin/django-admin.py checkmetarefresh --settings=peer.settings

This command looks for the entities that need to be refreshed and triggers the
update.

Domain Ownership Proof
~~~~~~~~~~~~~~~~~~~~~~

Currently only a mechanism to prove the ownership of a domain is implemented:
sending a specific HTTP request to a host on that domain. Other mechanism are
expected to appear in future releases.

Some web servers are configured to ban any request from an user agent that
they don't recognize. That is the reason there is a setting where you can
set a custom User Agent header to trick your web server into thinking this
requests does not come from a malicious bot.

.. code-block:: python

 DOP_USER_AGENT = 'Mozilla/5.0 (X11; Linux i686; rv:10.0.1) Gecko/20100101 Firefox/10.0.1'

This option is not set by default, which means the default user agent that
is used is specified in Python standard library. This happens to be
*Python-urllib/2.6*


Nagios Integration
~~~~~~~~~~~~~~~~~~

There is a config option to allow send events to nagios vía NSCA (nagios
agent) when a Entity is modified, createad or deleted.

Nagios hosts and services must be set to allow notifications for *service peer*
and *server entitydomain.fqdn* as configured in settings.

This a simple example service/hosts nagios config to allow NSCA:

.. code-block:: bash

 ;; Passive service template definition
 define service{
     name                    passive-service
     use                     generic-service
     check_freshness         1
     passive_checks_enabled  1
     active_checks_enabled   0
     is_volatile             0
     flap_detection_enabled  0
     notification_options    w,u,c,s
     freshness_threshold     57600     ;12hr
     check_command           check_dummy!3!"No Data Received"
 }

 define host{
     use             generic-host
     host_name       entitydomain.fqdn
     alias           testing peer domain
     address         192.168.1.122
     contact_groups  admins
 }

 define service{
     use                 passive-service
     host_name           entitydomain.fqdn
     service_description peer
     contact_groups      admins
 }


You must setup your nsca.conf on nagios server and send_nsca.conf on nagios
agent host.

*service_description* must be set as *NSCA_SERVICE* settings value
*host_name* is the same that domain field on Entity objects

Remeber that nsca agent must be installed on peer server.

To enable nagios notification from django peer application you must set the
correct properties on settings:

.. code-block:: python

  # Entities modificated nagios notification command (watch man send_nsca)
  # Disabled if None
  # NSCA_COMMAND = None
  NSCA_COMMAND = '/usr/sbin/send_nsca -H nagios.fqdn'

  # Nagios accept 0, 1, 2, 3 as 0=OK, 1=WARNING, 2=CRITICAL, 3=UNKNOWN
  NSCA_NOTIFICATION_LEVEL = 3


CHANGES.rst
CHANGES
=======

0.19.1 (2015-07-14)
-------------------

- Fix bug in entities URLs

0.19.0 (2015-07-03)
-------------------

- Changes in the home page
- Use the tables in the admin site for listings of entities
- Add table of text chunks for home page slogan, terms of use, etc.
- Users can choose role between entities operator and metadata registrar.

0.18.0 (2015-06-25)
-------------------

- Support for django 1.8

0.17.2 (2015-06-24)
-------------------

- Merge jkakavas' moderation fork

0.17.1 (2015-06-24)
-------------------

- Use the new built-in schema migration framework.

0.17.0 (2015-02-12)
-------------------

- Updated djangosaml2 to 0.13.0

0.16.0 (2015-02-04)
-------------------

- django 1.7 support
- some changes in the front page
- Updated GitPython to 0.3.6, gitdb to 0.6.4, django-vff to 0.2b3

0.14.0 (2014-11-20)
-------------------
- When a subdomain of a verified domain is added by the same user,
  automatically set it as verified.
- Add `publicsuffix <http://pypi.python.org/pypi/publicsuffix>`_ dependency.
- Check that newly added domains are valid.

0.13.0 (2014-02-27)
-------------------
- Upgrade to Django 1.6.2
- Upgrade djangosaml to 0.10.0
- Upgrade django-registration to 1.0
- Upgrade django-recaptcha to 0.0.9
- Adapt form fields to Django 1.6.2
- Adapt urls.py to Django 1.6.2
- Adapt url tag in templates to Django 1.6.2
- Fix license headers (80 char length)
- Add navigation menu to the header, for logged users

0.12.0 (2013-10-29)
-------------------
- Add swig dependence.
- Bugfix in 0007 migration.
- Display validation errors and warnings in a more useful way, with only
  relevant information.
- Fix some encoding errors.
- Support domain validation through email. Add python-whois dependence.
- Add support for domain validation through https.
- Notify the domain owner of validation activity related to that domain.
- Support domain invalidation.
- Bugfix in email notifications.
- Warn about nxdomain response cached in DNS-based validation.
- Validate that DiscoHints elements only appears on IdP SSO nodes.
- Fix parsing the mdUI Info.
- Enforce that KeyDescriptor nodes appear before NameIDFormat ones.
- Avoid crashing in the prettify function with comments.
- Catch the XMLSyntax exception to avoid 500 error when invalid metadata.
- Check for multiple SSO of the same type and warn the user, because that is
  unsupported by SAMLmetaJS.

0.11.0 (2012-04-18)
-------------------
- Add support for schema validation (code stolen from pyff).
- Allow more customizations in the federated login UI.

0.10.2 (2013-02-28)
-------------------
- Upgrade django-recaptcha so it supports SSL
- Minor cleanups of staled files
- Use just the hostname to validate if an endpoint belongs to a domain.
  This allows using non standard ports for the endpoints. Patch by
  Hans Zandbelt

0.10.1 (2012-07-29)
-------------------
- Fix url hiding by reordering some url definitions.
- When importing an entity's metadata inside the EntitiesDescriptor tag, strip
  off such tag.
- Update SAMLmeteaJS to latest version upstream to fix issue #3.
- Several fixes in the documentation:

  - Make the paths more generic and add notes with real paths depending on the
    current version.
  - Add a note about the initial administrator's name.
  - Replace terena.org with example.com in the example configuration.
  - Fix some typos.

- Include txt templates in the distribution.

0.10.0 (2012-03-28)
-------------------
- Nagios integration by implementing a passive nagios agent that
  listen for Entity's updates (creations, updates and deletions)
  and send them to a Nagios server.
- Remove PEER names. If the DisplayName is present it is used as
  the label for the entity. Otherwise, the entity id is used and
  if neither of them exists, the PEER numeric id is used.
- Improve the usability of the SAMLmetaJS editor by showing
  user errors as soon as possible and asking the user either to
  fix them or to remove the information that is giving trouble.
- Implement REMOTE_USER authentication with an easy to switch
  settings option. Also document the web server part of the setup.
- Update djangosaml2 dependency version. This uses newest pysaml2 version.
  As a consequence encrypted assertions and signed response and requests
  work better now. Djangosaml2 0.4.2 also adds logging support.

0.9.0 (2012-03-03)
------------------
- Several changes to the SAMLmetaJS editor:
  - Support for IdP endpoints and certificates.
  - Support for EncryptionMethod in KeyDescriptors
  - Support for MDUI.PrivacyStatementURL and MDUI.InformationURL.
  - Many refactorings to improve code reuse.
  - Merge features from PEER 0.8.0 into SAMLmetaJS master branch.
- Show IdP information (endpoints and certificates) in the details view.
- Improve certificate rendering in the details view.
- Send emails to subscribers of entities when their endpoints are down.
- Allow to subscribe to entities for updates about their endpoints state.
- Refactor the views module of the entities application into a package
  which is easier to handle.
- Several bug fixes:
  - Fix the terms of use file in the metadata upload file form (issue #2).
  - Fix bad use of ugettext_lazy and use ugettext instead.
  - Fix a crash when an entity is new and does not have metadata yet.
  - Fix the way the diff is computed when submitting a change to the
  metadata of an entity.

0.8.0 (2012-01-31)
------------------
- Rich metadata support. Now it is possible to edit the logo and
  geolocalization hint of an entity. This feature is used
  to display richer multimedia information through PEER.
- Improve the metadata editor with request initiator and
  discovery response endpoint support. Also improve the localization
  plugin and add keywords and logo to the information plugin.
- Metadata refresh: if the entity id is a URL that points to the
  metadata itself the user can activate a periodic fetch of this
  metadata.

peer/settings.py
# Copyright 2011 Terena. All rights reserved.

# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
#    1. Redistributions of source code must retain the above copyright notice,
#       this list of conditions and the following disclaimer.
#
#    2. Redistributions in binary form must reproduce the above copyright
#       notice, this list of conditions and the following disclaimer in the
#       documentation and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY TERENA ``AS IS'' AND ANY EXPRESS OR IMPLIED
# WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
# EVENT SHALL TERENA OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA,
# OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
# LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
# NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,
# EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
# The views and conclusions contained in the software and documentation are
# those of the authors and should not be interpreted as representing official
# policies, either expressed or implied, of Terena.

# Django settings for peer project.

import datetime
import os

_ = lambda s: s

BASEDIR = os.path.abspath(os.path.dirname(__file__))

DEBUG = False
TEMPLATE_DEBUG = DEBUG

ADMINS = (
    # ('Your Name', 'your_email@example.com'),
)

MANAGERS = ADMINS

DEFAULT_FROM_EMAIL = 'no-reply@example.com'

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.postgresql_psycopg2',  # Add 'postgresql_psycopg2', 'postgresql', 'mysql', 'sqlite3' or 'oracle'.
        'NAME': 'peer',                                      # Or path to database file if using sqlite3.
        'USER': 'peer',                                      # Not used with sqlite3.
        'PASSWORD': 'peer',                                  # Not used with sqlite3.
        'HOST': '',                                          # Set to empty string for localhost. Not used with sqlite3.
        'PORT': '',                                          # Set to empty string for default. Not used with sqlite3.
    }
}

# Local time zone for this installation. Choices can be found here:
# http://en.wikipedia.org/wiki/List_of_tz_zones_by_name
# although not all choices may be available on all operating systems.
# On Unix systems, a value of None will cause Django to use the same
# timezone as the operating system.
# If running in a Windows environment this must be set to the same as your
# system time zone.
TIME_ZONE = 'America/Chicago'

# Language code for this installation. All choices can be found here:
# http://www.i18nguy.com/unicode/language-identifiers.html
LANGUAGE_CODE = 'en-us'

SITE_ID = 1

# If you set this to False, Django will make some optimizations so as not
# to load the internationalization machinery.
USE_I18N = True

# If you set this to False, Django will not format dates, numbers and
# calendars according to the current locale
USE_L10N = True

# Absolute filesystem path to the directory that will hold user-uploaded files.
# Example: "/home/media/media.lawrence.com/media/"
MEDIA_ROOT = os.path.join(BASEDIR, 'media')

# URL that handles the media served from MEDIA_ROOT. Make sure to use a
# trailing slash.
# Examples: "http://media.lawrence.com/media/", "http://example.com/media/"
MEDIA_URL = '/media/'

# Absolute path to the directory static files should be collected to.
# Don't put anything in this directory yourself; store your static files
# in apps' "static/" subdirectories and in STATICFILES_DIRS.
# Example: "/home/media/media.lawrence.com/static/"
STATIC_ROOT = os.path.join(BASEDIR, 'static')

# URL prefix for static files.
# Example: "http://media.lawrence.com/static/"
STATIC_URL = '/static/'

# URL prefix for admin static files -- CSS, JavaScript and images.
# Make sure to use a trailing slash.
# Examples: "http://foo.com/static/admin/", "/static/admin/".
ADMIN_MEDIA_PREFIX = '/static/admin/'

# Additional locations of static files
STATICFILES_DIRS = (
    # Put strings here, like "/home/html/static" or "C:/www/django/static".
    # Always use forward slashes, even on Windows.
    # Don't forget to use absolute paths, not relative paths.
    os.path.join(BASEDIR, 'staticfiles'),
)

# Aditional theme custom styles
PEER_THEME = {
    'LINK_COLOR': '#5669CE',
    'LINK_HOVER': '#1631BC',
    'HEADING_COLOR': '#1631BC',
    'INDEX_HEADING_COLOR': '#ff7b33',
    'HEADER_BACKGROUND': '',
    'CONTENT_BACKGROUND': '',
    'FOOTER_BACKGROUND': '',
    'JQUERY_UI_THEME': 'default-theme',
}

# List of finder classes that know how to find static files in
# various locations.
STATICFILES_FINDERS = (
    'django.contrib.staticfiles.finders.FileSystemFinder',
    'django.contrib.staticfiles.finders.AppDirectoriesFinder',
    # 'django.contrib.staticfiles.finders.DefaultStorageFinder',
)

# Make this unique, and don't share it with anybody.
SECRET_KEY = 'w^e-s0n)+#^ottb%lm+%9=_etmv8p&s3hacm6^i41+1+u_nw-3'

# List of callables that know how to import templates from various sources.
TEMPLATE_LOADERS = (
    'django.template.loaders.filesystem.Loader',
    'django.template.loaders.app_directories.Loader',
    # 'django.template.loaders.eggs.Loader',
)


# List of processors used by RequestContext to populate the context.
TEMPLATE_CONTEXT_PROCESSORS = (
    'django.contrib.auth.context_processors.auth',
    'django.core.context_processors.debug',
    'django.core.context_processors.i18n',
    'django.core.context_processors.media',
    'django.core.context_processors.static',
    'django.core.context_processors.request',
    'django.contrib.messages.context_processors.messages',
    'peer.portal.context_processors.peer_theme',
    'peer.portal.context_processors.auth',
    'peer.entity.context_processors.get_settings',
)

MIDDLEWARE_CLASSES = (
    'django.middleware.common.CommonMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
)

ROOT_URLCONF = 'peer.urls'

TEMPLATE_DIRS = (
    # Put strings here, like "/home/html/django_templates" or
    # "C:/www/django/templates".
    # Always use forward slashes, even on Windows.
    # Don't forget to use absolute paths, not relative paths.
    os.path.join(BASEDIR, 'templates'),
)

INSTALLED_APPS = (
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.sites',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    'django.contrib.admin.apps.SimpleAdminConfig',
    'captcha',
    'registration',
    'djangosaml2',
    'peer.account',
    'peer.domain',
    'peer.entity',
    'peer.portal',
    'ckeditor',
)

# needed for django-registration
ACCOUNT_ACTIVATION_DAYS = 2
LOGIN_REDIRECT_URL = '/'

CKEDITOR_UPLOAD_PATH = "uploads/"

AUTHENTICATION_BACKENDS = (
    'djangosaml2.backends.Saml2Backend',
    'django.contrib.auth.backends.RemoteUserBackend',
    'django.contrib.auth.backends.ModelBackend',
)

# Email settings. Commented by security, but should be rewrite on a local or
# production settings
# EMAIL_HOST = 'smtp.example.com'
# EMAIL_PORT = 25

# reCaptcha keys. Not filled for security
RECAPTCHA_PUBLIC_KEY = ''
RECAPTCHA_PRIVATE_KEY = ''
# Uncomment the following line to make reCaptcha validation submits
# to be made over SSL
# RECAPTCHA_USE_SSL = True

# A sample logging configuration. The only tangible logging
# performed by this configuration is to send an email to
# the site admins on every HTTP 500 error.
# See http://docs.djangoproject.com/en/dev/topics/logging for
# more details on how to customize your logging configuration.
LOGGING = {
    'version': 1,
    'disable_existing_loggers': False,
    'handlers': {
        'mail_admins': {
            'level': 'ERROR',
            'class': 'django.utils.log.AdminEmailHandler'
        }
    },
    'loggers': {
        'django.request': {
            'handlers': ['mail_admins'],
            'level': 'ERROR',
            'propagate': True,
        },
        'peer.nagios': {
            'handlers': ['mail_admins'],
            'level': 'ERROR',
            'propagate': True,
        },
    }
}

# ENTITY METADATA VALIDATION
METADATA_VALIDATORS = (
    'peer.entity.validation.validate_xml_syntax',
    'peer.entity.validation.validate_domain_in_endpoints',
    'peer.entity.validation.validate_metadata_permissions',
    # 'peer.entity.validation.validate_domain_in_entityid',
    'peer.entity.validation.validate_schema',
)

# Permissions for metadata attribute. The metadata attribute is defined by
# its XPATH. The triplets are (<XPATH>, <Permission name>, <Permission
# description>)
#METADATA_PERMISSIONS = (
    #('/md:EntityDescriptor', 'entity_descriptor', 'Entity Descriptor'),
    #('.//md:ServiceDescription', 'service_descriptor', 'Service Description'),
    #('.//mdui:Description', 'description', 'Description'),
    #('.//md:OrganizationName', 'organization_name', 'Organization Name'),
#)

# VFF VERSIONED FILE BACKEND:
VFF_BACKEND = 'vff.git_backend.GitBackend'
# VFF (git backend) Path to the root of the git repo:
# VFF_REPO_ROOT = '/path/to/git/repo'
# VFF (git backend) Relative path within the git repo
# to the directory where vff keeps its managed files:
# VFF_REPO_PATH = 'my/dir'


# DOMAIN Verification by MAIL: Administrative email addresses supported
ADMINISTRATIVE_EMAIL_ADDRESSES = ('webmaster', 'hostmaster', 'abuse',
                                  'postmaster', 'admin')

# DOMAIN Invalidation: Send to the administrative contact of a domain
# returned by a WHOIS LOOKUP an email when the domain is validated to allow
# revoke the action
NOTIFY_DOMAIN_OWNER = False

# SAMLmetaJS plugins
SAML_META_JS_PLUGINS = ('info', 'org', 'contact', 'endpoints', 'certs',
                        'attributes', 'entityattrs', 'location')


# Terms of Use
USER_REGISTER_TERMS_OF_USE = os.path.join(BASEDIR,
                                          'user_register_terms_of_use.txt')
METADATA_IMPORT_TERMS_OF_USE = os.path.join(BASEDIR,
                                            'metadata_import_terms_of_use.txt')

# Max number of entries the global RSS feed will return
# MAX_FEED_ENTRIES = 10

# Expiration warnings
EXPIRATION_WARNING_TIMEDELTA = datetime.timedelta(days=1)

# Entities pagination
ENTITIES_PER_PAGE = 10

# Entities modificated nagios notification command (watch man send_nsca)
# Disabled if None
# NSCA_COMMAND = '/usr/sbin/send_nsca -H nagios.fqdn'
NSCA_COMMAND = None

# Nagios accept 0, 1, 2, 3 as 0=OK, 1=WARNING, 2=CRITICAL, 3=UNKNOWN
NSCA_NOTIFICATION_LEVEL = 3

# Nagios service name
NSCA_SERVICE = 'peer'

# Federated auth
SAML_ENABLED = False
SESSION_EXPIRE_AT_BROWSER_CLOSE = SAML_ENABLED
SAML_CREATE_UNKNOWN_USER = True
SAML_ATTRIBUTE_MAPPING = {
    'mail': ('username', 'email'),
    'givenName': ('first_name', ),
    'sn': ('last_name', ),
}

SAML_ONE_IDP_SIGN_IN_BUTTON = _('Federated sign in')
SAML_SEVERAL_IDPS_SIGN_IN_BUTTON = _('Federated sign in')

SAML_CONFIG = {}  # YOU MUST OVERWRITE THIS IN THE LOCAL SETTINGS

# Remote user auth
REMOTE_USER_ENABLED = False

# Moderation
MODERATION_ENABLED = True

try:
    from local_settings import *
except ImportError:
    pass
